layers: Reduce looping for builtins

sfricke-samsung opened the pull request:
Related to https://github.com/KhronosGroup/Vulkan-ValidationLayers/projects/4#card-50486050.

Few spots were looping the first part of the shader just to find the built-in information. Moved that to the initial pass since we are already collecting the decorations there for other reasons

----

(note: this will totally merge conflict in the head file with !2596 and I will fix the later, in case you want to save yourself a internal CI run on this until the other PR is in)

ncesario-lunarg commented on layers/shader_validation.h:
```diff
@@ -129,6 +129,11 @@ struct function_set {
 
+struct builtin_set {
+    spirv_inst_iter insn;  // OpDecorate or OpMemberDecorate
+    uint32_t builtin;      // spv::BuiltIn
```
Should this have type `spv::BuiltIn` if that's what it represents?

sfricke-samsung replied:
I think your right, it would be more clear then having a comment... will update

ncesario-lunarg reviewed (approved):
LGTM! Minor nit about the `builtin_set::builtin` type, but I might be missing something obvious there.

ncesario-lunarg commented:
Running through [internal CI](http://erusea:4040/8513).

sfricke-samsung commented:
@ncesario-lunarg rebased, updated `builtin_set` struct, and fixed a few spots where the code was using `built_in` and changed to `builtin` to keep naming consistent

ncesario-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -184,13 +184,20 @@ void SHADER_MODULE_STATE::BuildDefIndex() {
                 decorations[target_id].add(insn.word(2), insn.len() > 3u ? insn.word(3) : 0u);
                 decoration_inst.push_back(insn);
+                if (insn.word(2) == spv::DecorationBuiltIn) {
+                    builtin_decoration_list.emplace_back(builtin_set{insn.offset(), static_cast<spv::BuiltIn>(insn.word(3))});
```
I _think_ having the constructor called in here will defeat the purpose of using emplace. e.g., to take advantage of emplace I believe you need to add a constructor to `builtin_set` and do something like:
```cpp
builtin_decoration_list.emplace_back(insn.offset(), static_cast<spv::BuiltIn>(insn.word(3)));
```

I think what's currently here will call the constructor followed by the move constructor (e.g., [here](https://godbolt.org/z/cGsW59)). Not a big deal here one way or the other, though.

sfricke-samsung replied:
no your right, this seems wrong, I think I had a reason before, but there was much refactoring since then as well... let me try cleaning this up

sfricke-samsung replied:
So with `builtin_decoration_list.emplace_back(insn.offset(), static_cast<spv::BuiltIn>(insn.word(3)));` you still need to add an explicit constructor `buildin_set(offset, builtin) : offset(offset), builtin(bulitin);` to the struct 

I guess maybe `builtin_decoration_list.push_back(builtin_set{insn.offset(), static_cast<spv::BuiltIn>(insn.word(3))});` is what we want instead.

ncesario-lunarg commented:
:+1: Re-running through CI; will merge once that finishes.
